File: app/src/main/java/eu/darken/bluemusic/main/ui/managed/ManagedDevicesAdapter.java
package eu.darken.bluemusic.main.ui.managed;

import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.annotation.RequiresApi;
import butterknife.BindView;
import eu.darken.bluemusic.R;
import eu.darken.bluemusic.main.core.audio.AudioStream;
import eu.darken.bluemusic.main.core.database.ManagedDevice;
import eu.darken.bluemusic.util.ApiHelper;
import eu.darken.bluemusic.util.AppTool;
import eu.darken.bluemusic.util.Check;
import eu.darken.bluemusic.util.DeviceHelper;
import eu.darken.bluemusic.util.ui.BasicAdapter;
import eu.darken.bluemusic.util.ui.BasicViewHolder;
import timber.log.Timber;

import static android.graphics.Typeface.BOLD;
import static android.graphics.Typeface.NORMAL;


class ManagedDevicesAdapter extends BasicAdapter<ManagedDevicesAdapter.ManagedDeviceVH, ManagedDevice> {
    private final Callback callback;

    ManagedDevicesAdapter(Callback callback) {
        this.callback = callback;
    }

    @Override
    public ManagedDeviceVH onCreateBaseViewHolder(LayoutInflater inflater, ViewGroup parent, int viewType) {
        return new ManagedDeviceVH(parent, callback);
    }

    interface Callback {
        void onStreamAdjusted(ManagedDevice device, AudioStream.Type type, float percentage);

        void onShowConfigScreen(ManagedDevice device);
    }

    static class ManagedDeviceVH extends BasicViewHolder<ManagedDevice> {
        @BindView(R.id.device_icon) ImageView icon;
        @BindView(R.id.name) TextView nameView;
        @BindView(R.id.lastseen) TextView lastSeen;

        @BindView(R.id.launch_icon) ImageView launchIcon;
        @BindView(R.id.autoplay_icon) ImageView autoPlayIcon;
        @BindView(R.id.volumelock_icon) ImageView volumeLockIcon;
        @BindView(R.id.keepawake_icon) ImageView keepAwakeIcon;
        @BindView(R.id.extras_container) ViewGroup extrasContainer;

        @BindView(R.id.config_icon) View config;

        @BindView(R.id.music_container) View musicContainer;
        @BindView(R.id.music_seekbar) SeekBar musicSeekbar;
        @BindView(R.id.music_counter) TextView musicCounter;

        @BindView(R.id.call_container) View voiceContainer;
        @BindView(R.id.call_seekbar) SeekBar voiceSeekbar;
        @BindView(R.id.call_counter) TextView voiceCounter;

        @BindView(R.id.ring_container) View ringContainer;
        @BindView(R.id.ring_seekbar) SeekBar ringSeekbar;
        @BindView(R.id.ring_counter) TextView ringCounter;
        @BindView(R.id.ring_permission_label) TextView ringPermissionLabel;
        @BindView(R.id.ring_permission_action) TextView ringPermissionAction;

        @BindView(R.id.notification_container) View notificationContainer;
        @BindView(R.id.notification_seekbar) SeekBar notificationSeekbar;
        @BindView(R.id.notification_counter) TextView notificationCounter;
        @BindView(R.id.notification_permission_label) TextView notificationPermissionLabel;
        @BindView(R.id.notification_permission_action) TextView notificationPermissionAction;
        private Callback callback;

        ManagedDeviceVH(ViewGroup parent, Callback callback) {
            super(parent, R.layout.viewholder_managed_device);
            this.callback = callback;
        }

        @RequiresApi(api = Build.VERSION_CODES.M)
        @Override
        public void bind(ManagedDevice item) {
            super.bind(item);
            icon.setImageResource(DeviceHelper.getIconForDevice(item.getSourceDevice()));

            nameView.setText(DeviceHelper.getAliasAndName(item.getSourceDevice()));
            nameView.setTypeface(null, item.isActive() ? BOLD : NORMAL);

            String timeString;
            if (item.getLastConnected() > 0) {
                timeString = DateUtils.getRelativeDateTimeString(getContext(), item.getLastConnected(), DateUtils.MINUTE_IN_MILLIS, DateUtils.WEEK_IN_MILLIS, 0).toString();
            } else {
                timeString = getString(R.string.label_neverseen);
            }
            lastSeen.setText(item.isActive() ? getString(R.string.label_state_connected) : timeString);

            autoPlayIcon.setVisibility(item.getAutoPlay() ? View.VISIBLE : View.GONE);

            volumeLockIcon.setVisibility(item.getVolumeLock() ? View.VISIBLE : View.GONE);
            keepAwakeIcon.setVisibility(item.getKeepAwake() ? View.VISIBLE : View.GONE);

            if (item.getLaunchPkg() != null) {
                try {
                    launchIcon.setImageDrawable(AppTool.getIcon(getContext(), item.getLaunchPkg()));
                } catch (PackageManager.NameNotFoundException e) {
                    Timber.e(e);
                }
            }
            launchIcon.setVisibility(item.getLaunchPkg() != null ? View.VISIBLE : View.GONE);

            extrasContainer.setVisibility(item.getLaunchPkg() != null || item.getAutoPlay() || item.getVolumeLock() ? View.VISIBLE : View.GONE);

            config.setOnClickListener(v -> callback.onShowConfigScreen(item));

            musicContainer.setVisibility(item.getVolume(AudioStream.Type.MUSIC) != null ? View.VISIBLE : View.GONE);
            if (item.getVolume(AudioStream.Type.MUSIC) != null) {
                musicSeekbar.setMax(item.getMaxVolume(AudioStream.Type.MUSIC));
                musicSeekbar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
                    @Override
                    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                        musicCounter.setText(String.valueOf(progress));
                    }

                    @Override
                    public void onStartTrackingTouch(SeekBar seekBar) {

                    }

                    @Override
                    public void onStopTrackingTouch(SeekBar seekBar) {
                        callback.onStreamAdjusted(item, AudioStream.Type.MUSIC, (float) seekBar.getProgress() / seekBar.getMax());
                    }
                });
                musicSeekbar.setProgress(item.getRealVolume(AudioStream.Type.MUSIC));
                musicCounter.setText(String.valueOf(musicSeekbar.getProgress()));
            }

            voiceContainer.setVisibility(item.getVolume(AudioStream.Type.CALL) != null ? View.VISIBLE : View.GONE);
            if (item.getVolume(AudioStream.Type.CALL) != null) {
                voiceSeekbar.setMax(item.getMaxVolume(AudioStream.Type.CALL));
                voiceSeekbar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
                    @Override
                    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                        voiceCounter.setText(String.valueOf(progress));
                    }

                    @Override
                    public void onStartTrackingTouch(SeekBar seekBar) {

                    }

                    @Override
                    public void onStopTrackingTouch(SeekBar seekBar) {
                        callback.onStreamAdjusted(item, AudioStream.Type.CALL, (float) seekBar.getProgress() / seekBar.getMax());
                    }
                });
                voiceSeekbar.setProgress(item.getRealVolume(AudioStream.Type.CALL));
                voiceCounter.setText(String.valueOf(voiceSeekbar.getProgress()));
            }

            ringContainer.setVisibility(item.getVolume(AudioStream.Type.RINGTONE) != null ? View.VISIBLE : View.GONE);
            if (item.getVolume(AudioStream.Type.RINGTONE) != null) {
                ringSeekbar.setMax(item.getMaxVolume(AudioStream.Type.RINGTONE));
                ringSeekbar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
                    @Override
                    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                        ringCounter.setText(String.valueOf(progress));
                    }

                    @Override
                    public void onStartTrackingTouch(SeekBar seekBar) {

                    }

                    @Override
                    public void onStopTrackingTouch(SeekBar seekBar) {
                        callback.onStreamAdjusted(item, AudioStream.Type.RINGTONE, (float) seekBar.getProgress() / seekBar.getMax());
                    }
                });
                ringSeekbar.setProgress(item.getRealVolume(AudioStream.Type.RINGTONE));
                ringCounter.setText(String.valueOf(ringSeekbar.getProgress()));
            }

            ringPermissionAction.setOnClickListener(v -> {
                Intent intent = new Intent(android.provider.Settings.ACTION_NOTIFICATION_POLICY_ACCESS_SETTINGS);
                getContext().startActivity(intent);
            });

            final NotificationManager notifMan = (NotificationManager) getContext().getSystemService(Context.NOTIFICATION_SERVICE);
            Check.notNull(notifMan);
            boolean needsPermission = ApiHelper.hasMarshmallow() && item.getVolume(AudioStream.Type.RINGTONE) != null && !notifMan.isNotificationPolicyAccessGranted();
            ringPermissionLabel.setVisibility(needsPermission ? View.VISIBLE : View.GONE);
            ringPermissionAction.setVisibility(needsPermission ? View.VISIBLE : View.GONE);
            ringSeekbar.setVisibility(needsPermission ? View.GONE : View.VISIBLE);
            ringCounter.setVisibility(needsPermission ? View.GONE : View.VISIBLE);


            notificationContainer.setVisibility(item.getVolume(AudioStream.Type.NOTIFICATION) != null ? View.VISIBLE : View.GONE);
            if (item.getVolume(AudioStream.Type.NOTIFICATION) != null) {
                notificationSeekbar.setMax(item.getMaxVolume(AudioStream.Type.NOTIFICATION));
                notificationSeekbar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
                    @Override
                    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                        notificationCounter.setText(String.valueOf(progress));
                    }

                    @Override
                    public void onStartTrackingTouch(SeekBar seekBar) {

                    }

                    @Override
                    public void onStopTrackingTouch(SeekBar seekBar) {
                        callback.onStreamAdjusted(item, AudioStream.Type.NOTIFICATION,
                                (float) seekBar.getProgress() / seekBar.getMax());
                    }
                });
                notificationSeekbar.setProgress(item.getRealVolume(AudioStream.Type.NOTIFICATION));
                notificationCounter.setText(String.valueOf(notificationSeekbar.getProgress()));
            }

            notificationPermissionAction.setOnClickListener(v -> {
                Intent intent = new Intent(android.provider.Settings.ACTION_NOTIFICATION_POLICY_ACCESS_SETTINGS);
                getContext().startActivity(intent);
            });

            notificationPermissionLabel.setVisibility(needsPermission ? View.VISIBLE : View.GONE);
            notificationPermissionAction.setVisibility(needsPermission ? View.VISIBLE : View.GONE);
            notificationSeekbar.setVisibility(needsPermission ? View.GONE : View.VISIBLE);
            notificationCounter.setVisibility(needsPermission ? View.GONE : View.VISIBLE);
        }
    }
}

File: app/src/main/java/eu/darken/bluemusic/main/ui/managed/ManagedDevicesAdapter.kt
package eu.darken.bluemusic.main.ui.managed

import android.app.Activity
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Typeface
import android.os.Build
import android.provider.Settings
import android.text.format.DateUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.annotation.RequiresApi
import eu.darken.bluemusic.R
import eu.darken.bluemusic.databinding.ViewholderManagedDeviceBinding
import eu.darken.bluemusic.main.core.audio.AudioStream
import eu.darken.bluemusic.main.core.database.ManagedDevice
import eu.darken.bluemusic.main.ui.managed.ManagedDevicesAdapter.ManagedDeviceVH
import eu.darken.bluemusic.util.*
import eu.darken.bluemusic.util.ui.BasicAdapter
import eu.darken.bluemusic.util.ui.BasicViewHolder
import timber.log.Timber

internal class ManagedDevicesAdapter(private val callback: Callback) : BasicAdapter<ManagedDeviceVH, ManagedDevice>() {
    override fun onCreateBaseViewHolder(inflater: LayoutInflater, parent: ViewGroup, viewType: Int): ManagedDeviceVH {
        return ManagedDeviceVH(parent, callback)
    }

    internal interface Callback {
        fun onStreamAdjusted(device: ManagedDevice, type: AudioStream.Type, percentage: Float)
        fun onShowConfigScreen(device: ManagedDevice)
    }

    internal class ManagedDeviceVH(parent: ViewGroup, private val callback: Callback) : BasicViewHolder<ManagedDevice>(parent, R.layout.viewholder_managed_device) {

        private val ui: ViewholderManagedDeviceBinding by lazy { ViewholderManagedDeviceBinding.bind(itemView) }

        @RequiresApi(api = Build.VERSION_CODES.M)
        override fun bind(item: ManagedDevice) {
            super.bind(item)
            ui.deviceIcon.setImageResource(DeviceHelper.getIconForDevice(item.sourceDevice))
            ui.name.text = DeviceHelper.getAliasAndName(item.sourceDevice)
            ui.name.setTypeface(null, if (item.isActive) Typeface.BOLD else Typeface.NORMAL)
            val timeString: String = if (item.lastConnected > 0) {
                DateUtils.getRelativeDateTimeString(context, item.lastConnected, DateUtils.MINUTE_IN_MILLIS, DateUtils.WEEK_IN_MILLIS, 0).toString()
            } else {
                getString(R.string.label_neverseen)
            }
            ui.lastseen.text = if (item.isActive) getString(R.string.label_state_connected) else timeString
            ui.lastseen.visibility = if (item.autoPlay) View.VISIBLE else View.GONE
            ui.volumelockIcon.visibility = if (item.volumeLock) View.VISIBLE else View.GONE
            ui.keepawakeIcon.visibility = if (item.keepAwake) View.VISIBLE else View.GONE
            if (item.launchPkg != null) {
                try {
                    ui.launchIcon.setImageDrawable(AppTool.getIcon(context, item.launchPkg))
                } catch (e: PackageManager.NameNotFoundException) {
                    Timber.e(e)
                }
            }
            ui.launchIcon.visibility = if (item.launchPkg != null) View.VISIBLE else View.GONE
            val hasExtras = item.launchPkg != null || item.autoPlay || item.volumeLock || item.keepAwake
            ui.extrasContainer.visibility = if (hasExtras) View.VISIBLE else View.GONE
            ui.configIcon.setOnClickListener { v: View? -> callback.onShowConfigScreen(item) }
            ui.musicContainer.visibility = if (item.getVolume(AudioStream.Type.MUSIC) != null) View.VISIBLE else View.GONE
            if (item.getVolume(AudioStream.Type.MUSIC) != null) {
                ui.musicSeekbar.max = item.getMaxVolume(AudioStream.Type.MUSIC)
                ui.musicSeekbar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
                    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                        ui.musicCounter.text = progress.toString()
                    }

                    override fun onStartTrackingTouch(seekBar: SeekBar) {}
                    override fun onStopTrackingTouch(seekBar: SeekBar) {
                        callback.onStreamAdjusted(item, AudioStream.Type.MUSIC, seekBar.progress.toFloat() / seekBar.max)
                    }
                })
                ui.musicSeekbar.progress = item.getRealVolume(AudioStream.Type.MUSIC)
                ui.musicCounter.text = ui.musicSeekbar.progress.toString()
            }
            ui.callContainer.visibility = if (item.getVolume(AudioStream.Type.CALL) != null) View.VISIBLE else View.GONE
            if (item.getVolume(AudioStream.Type.CALL) != null) {
                ui.callSeekbar.max = item.getMaxVolume(AudioStream.Type.CALL)
                ui.callSeekbar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
                    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                        ui.callCounter.text = progress.toString()
                    }

                    override fun onStartTrackingTouch(seekBar: SeekBar) {}
                    override fun onStopTrackingTouch(seekBar: SeekBar) {
                        callback.onStreamAdjusted(item, AudioStream.Type.CALL, seekBar.progress.toFloat() / seekBar.max)
                    }
                })
                ui.callSeekbar.progress = item.getRealVolume(AudioStream.Type.CALL)
                ui.callCounter.text = ui.callSeekbar.progress.toString()
            }
            ui.ringContainer.visibility = if (item.getVolume(AudioStream.Type.RINGTONE) != null) View.VISIBLE else View.GONE
            if (item.getVolume(AudioStream.Type.RINGTONE) != null) {
                ui.ringSeekbar.max = item.getMaxVolume(AudioStream.Type.RINGTONE)
                ui.ringSeekbar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
                    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                        ui.ringCounter.text = progress.toString()
                    }

                    override fun onStartTrackingTouch(seekBar: SeekBar) {}
                    override fun onStopTrackingTouch(seekBar: SeekBar) {
                        callback.onStreamAdjusted(item, AudioStream.Type.RINGTONE, seekBar.progress.toFloat() / seekBar.max)
                    }
                })
                ui.ringSeekbar.progress = item.getRealVolume(AudioStream.Type.RINGTONE)
                ui.ringCounter.text = ui.ringSeekbar.progress.toString()
            }
            ui.ringPermissionAction.setOnClickListener { v: View? -> ActivityUtil.tryStartActivity(context as Activity, Intent(Settings.ACTION_NOTIFICATION_POLICY_ACCESS_SETTINGS)) }
            val notifMan = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            Check.notNull(notifMan)
            val needsPermission = ApiHelper.hasMarshmallow() && item.getVolume(AudioStream.Type.RINGTONE) != null && !notifMan.isNotificationPolicyAccessGranted
            ui.ringPermissionLabel.visibility = if (needsPermission) View.VISIBLE else View.GONE
            ui.ringPermissionAction.visibility = if (needsPermission) View.VISIBLE else View.GONE
            ui.ringSeekbar.visibility = if (needsPermission) View.GONE else View.VISIBLE
            ui.ringCounter.visibility = if (needsPermission) View.GONE else View.VISIBLE
            ui.notificationContainer.visibility = if (item.getVolume(AudioStream.Type.NOTIFICATION) != null) View.VISIBLE else View.GONE
            if (item.getVolume(AudioStream.Type.NOTIFICATION) != null) {
                ui.notificationSeekbar.max = item.getMaxVolume(AudioStream.Type.NOTIFICATION)
                ui.notificationSeekbar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
                    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                        ui.notificationCounter.text = progress.toString()
                    }

                    override fun onStartTrackingTouch(seekBar: SeekBar) {}
                    override fun onStopTrackingTouch(seekBar: SeekBar) {
                        callback.onStreamAdjusted(item, AudioStream.Type.NOTIFICATION,
                                seekBar.progress.toFloat() / seekBar.max)
                    }
                })
                ui.notificationSeekbar.progress = item.getRealVolume(AudioStream.Type.NOTIFICATION)
                ui.notificationCounter.text = ui.notificationSeekbar.progress.toString()
            }
            ui.notificationPermissionAction.setOnClickListener { v: View? -> ActivityUtil.tryStartActivity(context as Activity, Intent(Settings.ACTION_NOTIFICATION_POLICY_ACCESS_SETTINGS)) }
            ui.notificationPermissionLabel.visibility = if (needsPermission) View.VISIBLE else View.GONE
            ui.notificationPermissionAction.visibility = if (needsPermission) View.VISIBLE else View.GONE
            ui.notificationSeekbar.visibility = if (needsPermission) View.GONE else View.VISIBLE
            ui.notificationCounter.visibility = if (needsPermission) View.GONE else View.VISIBLE
            ui.alarmContainer.visibility = if (item.getVolume(AudioStream.Type.ALARM) != null) View.VISIBLE else View.GONE
            if (item.getVolume(AudioStream.Type.ALARM) != null) {
                ui.alarmSeekbar.max = item.getMaxVolume(AudioStream.Type.ALARM)
                ui.alarmSeekbar.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
                    override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                        ui.alarmCounter.text = progress.toString()
                    }

                    override fun onStartTrackingTouch(seekBar: SeekBar) {}
                    override fun onStopTrackingTouch(seekBar: SeekBar) {
                        callback.onStreamAdjusted(item, AudioStream.Type.ALARM, seekBar.progress.toFloat() / seekBar.max)
                    }
                })
                ui.alarmSeekbar.progress = item.getRealVolume(AudioStream.Type.ALARM)
                ui.alarmCounter.text = ui.alarmSeekbar.progress.toString()
            }
        }
    }
}